src/drivers/rest/data.rs

```diff
@@ -0,0 +1,165 @@
+    pub message: T,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct MAVLinkMessageStatus {
+    message: mavlink::ardupilotmega::MavMessage,
+    status: Status,
+}
+
+impl MAVLinkMessageStatus {
+    fn update(&mut self, message: &MAVLinkMessage<mavlink::ardupilotmega::MavMessage>) {
+        self.message = message.message.clone();
+        self.status.update();
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct MAVLinkVehicleComponentData {
+    id: u8,
+    messages: HashMap<String, MAVLinkMessageStatus>,
+}
+
+impl MAVLinkVehicleComponentData {
+    fn update(&mut self, message: &MAVLinkMessage<mavlink::ardupilotmega::MavMessage>) {
+        let message_name = message.message.message_name().to_string();
+        self.messages
+            .entry(message_name)
+            .or_insert(MAVLinkMessageStatus {
+                message: message.message.clone(),
+                status: Status::default(),
+            })
+            .update(message);
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct MAVLinkVehicleData {
+    id: u8,
+    components: HashMap<u8, MAVLinkVehicleComponentData>,
+}
+
+impl MAVLinkVehicleData {
+    fn update(&mut self, message: &MAVLinkMessage<mavlink::ardupilotmega::MavMessage>) {
+        let component_id = message.header.component_id;
+        self.components
+            .entry(component_id)
+            .or_insert(MAVLinkVehicleComponentData {
+                id: component_id,
+                messages: HashMap::new(),
+            })
+            .update(message);
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct MAVLinkVehiclesData {
+    vehicles: HashMap<u8, MAVLinkVehicleData>,
+}
+
+impl MAVLinkVehiclesData {
+    //TODO: Move message to reference
+    fn update(&mut self, message: MAVLinkMessage<mavlink::ardupilotmega::MavMessage>) {
+        let vehicle_id = message.header.system_id;
+        self.vehicles
+            .entry(vehicle_id)
+            .or_insert(MAVLinkVehicleData {
+                id: vehicle_id,
+                components: HashMap::new(),
+            })
+            .update(&message);
+    }
+
+    pub fn pointer(&self, path: &str) -> String {
+        if path.is_empty() {
+            return serde_json::to_string_pretty(self).unwrap();
+        }
+
+        let path = format!("/{path}");
+
+        dbg!(&path);
+
+        if path == "/vehicles" {
+            return serde_json::to_string_pretty(&self.vehicles).unwrap();
+        };
+
+        let value = serde_json::to_value(self).unwrap();
+        return match value.pointer(&path) {
+            Some(content) => serde_json::to_string_pretty(content).unwrap(),
+            None => "None".into(),
+        };
+    }
+}
+
+#[derive(Debug)]
+struct Data {
+    messages: Arc<Mutex<MAVLinkVehiclesData>>,
+}
+
+lazy_static! {
+    static ref DATA: Data = Data {
+        messages: Arc::new(Mutex::new(MAVLinkVehiclesData::default())),
+    };
+}
```

Comment:
Can you change the order of all these declarations to be top-to-down and not down-to-top? Like:

```Rust
use std::sync::{Arc, Mutex};

lazy_static!{
    static ref AAAA: A: A {...}
}

A {b: Arc<Mutex<<B>>}

impl A {...}

B {c: Vec<C>}

impl B {...}

C {...}

impl C {...}

fn freefunction() {...}
```

Reply:
While I was doing the rework.. I took a look in [gstreamer](https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/blob/main/tutorial/src/rgb2gray/imp.rs?ref_type=heads) and [egui](https://docs.rs/egui/latest/src/egui/menu.rs.html). Both follow down-to-top pattern.

Also, [tokio](https://github.com/tokio-rs/tokio/blob/2c14f88c90947b5fd845892b3d968f7b342db04e/tokio/src/runtime/context/current.rs) uses declaration first over implementation.
```
struct A;
struct B;

impl A;
impl B;
```

I failed to found top-to-down examples under tokio, chrono, std, egui and gstreamer. From my conclusion, this appears to be an anti-pattern now. We should move to "down-to-top". There is no reason to be agains the biggest part of the community.

Reply:
The impl vs declaration is not the point. I indeed prefer all declaration first, then all implementations. Apmost all crates use the top-to-down declarations - take a look at the order in which the interdependency of the declarations follow.

Reply:
In the case of keeping all decl first while using top-to-down:

```rust
use std::sync::{Arc, Mutex};

lazy_static!{
    static ref AAAA: A: A {...}
}

A {b: Arc<Mutex<<B>>}

B {c: Vec<C>}

C {...}

impl A {...}

impl B {...}

impl C {...}

fn freefunction() {...}
```

Reply:
> Apmost all crates use the top-to-down declarations

Did you check my links ? I was unable to find references for such thing. It appears to be not common. Which crates use such approach ?

Reply:
Yes, I did, and they proof exaclty what I am stating: top-to-down, regardless of doing <decl,impl,decl,impl> or <decl,decl,imp,imp>:

[gstreamer](https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/blob/main/tutorial/src/rgb2gray/imp.rs?ref_type=heads):
```rust
// This module contains the private implementation details of our element
//
static CAT: Lazy<gst::DebugCategory> = Lazy::new(|| {
    gst::DebugCategory::new(
        "rsrgb2gray",
        gst::DebugColorFlags::empty(),
        Some("Rust RGB-GRAY converter"),
    )
});

// Default values of properties
const DEFAULT_INVERT: bool = false;
const DEFAULT_SHIFT: u32 = 0;

// Property value storage
#[derive(Debug, Clone, Copy)]
struct Settings {
    invert: bool,
    shift: u32,
}

impl Default for Settings {
    fn default() -> Self {
        Settings {
            invert: DEFAULT_INVERT,
            shift: DEFAULT_SHIFT,
        }
    }
}

// Struct containing all the element data
#[derive(Default)]
pub struct Rgb2Gray {
    settings: Mutex<Settings>,
}

impl Rgb2Gray {
    // Converts one pixel of BGRx to a grayscale value, shifting and/or
    // inverting it as configured
    #[inline]
    fn bgrx_to_gray(in_p: &[u8], shift: u8, invert: bool) -> u8 {
        // See https://en.wikipedia.org/wiki/YUV#SDTV_with_BT.601
        const R_Y: u32 = 19595; // 0.299 * 65536
        const G_Y: u32 = 38470; // 0.587 * 65536
        const B_Y: u32 = 7471; // 0.114 * 65536

        assert_eq!(in_p.len(), 4);

        let b = u32::from(in_p[0]);
        let g = u32::from(in_p[1]);
        let r = u32::from(in_p[2]);

        let gray = ((r * R_Y) + (g * G_Y) + (b * B_Y)) / 65536;
        let gray = (gray as u8).wrapping_add(shift);

        if invert {
            255 - gray
        } else {
            gray
        }
    }
}

```

[egui](https://docs.rs/egui/latest/src/egui/menu.rs.html#293-295):
```rust
/// What is saved between frames.
#[derive(Clone, Default)]
pub struct [BarState](https://docs.rs/egui/latest/egui/menu/struct.BarState.html) {
    open_menu: [MenuRootManager](https://docs.rs/egui/latest/src/egui/menu.rs.html#293-295),
}

impl [BarState](https://docs.rs/egui/latest/src/egui/menu.rs.html#33-35) {
    pub fn load(ctx: &[Context](https://docs.rs/egui/latest/src/egui/context.rs.html#724), bar_id: [Id](https://docs.rs/egui/latest/src/egui/id.rs.html#34)) -> Self {
        [ctx](https://docs.rs/egui/latest/src/egui/menu.rs.html#38).[data_mut](https://docs.rs/egui/latest/src/egui/context.rs.html#941-943)(|d| [d](https://docs.rs/egui/latest/src/egui/menu.rs.html#39).[get_temp](https://docs.rs/egui/latest/src/egui/util/id_type_map.rs.html#385-388)::<Self>([bar_id](https://docs.rs/egui/latest/src/egui/menu.rs.html#38)).[unwrap_or_default](https://doc.rust-lang.org/nightly/core/option/enum.Option.html#method.unwrap_or_default)())
    }

    pub fn store(self, ctx: &[Context](https://docs.rs/egui/latest/src/egui/context.rs.html#724), bar_id: [Id](https://docs.rs/egui/latest/src/egui/id.rs.html#34)) {
        [ctx](https://docs.rs/egui/latest/src/egui/menu.rs.html#42).[data_mut](https://docs.rs/egui/latest/src/egui/context.rs.html#941-943)(|d| [d](https://docs.rs/egui/latest/src/egui/menu.rs.html#43).[insert_temp](https://docs.rs/egui/latest/src/egui/util/id_type_map.rs.html#369-372)([bar_id](https://docs.rs/egui/latest/src/egui/menu.rs.html#42), [self](https://docs.rs/egui/latest/src/egui/menu.rs.html#42)));
    }

    /// Show a menu at pointer if primary-clicked response.
    ///
    /// Should be called from [`Context`] on a [`Response`]
    pub fn bar_menu<R>(
        &mut self,
        button: &[Response](https://docs.rs/egui/latest/src/egui/response.rs.html#22-136),
        add_contents: impl FnOnce(&mut [Ui](https://docs.rs/egui/latest/src/egui/ui.rs.html#51-113)) -> R,
    ) -> Option<InnerResponse<R>> {
        [MenuRoot::stationary_click_interaction](https://docs.rs/egui/latest/src/egui/menu.rs.html#480-483)([button](https://docs.rs/egui/latest/src/egui/menu.rs.html#51), &mut [self](https://docs.rs/egui/latest/src/egui/menu.rs.html#50).open_menu);
        [self](https://docs.rs/egui/latest/src/egui/menu.rs.html#50).open_menu.[show](https://docs.rs/egui/latest/src/egui/menu.rs.html#301-315)([button](https://docs.rs/egui/latest/src/egui/menu.rs.html#51), [add_contents](https://docs.rs/egui/latest/src/egui/menu.rs.html#52))
    }

    pub(crate) fn has_root(&self) -> [bool](https://doc.rust-lang.org/nightly/std/primitive.bool.html) {
        [self](https://docs.rs/egui/latest/src/egui/menu.rs.html#58).open_menu.inner.[is_some](https://doc.rust-lang.org/nightly/core/option/enum.Option.html#method.is_some)()
    }
}

impl [std::ops::Deref](https://doc.rust-lang.org/nightly/core/ops/deref/trait.Deref.html) for [BarState](https://docs.rs/egui/latest/src/egui/menu.rs.html#33-35) {
    type Target = [MenuRootManager](https://docs.rs/egui/latest/src/egui/menu.rs.html#293-295);

    fn deref(&self) -> &Self::Target {
        &[self](https://docs.rs/egui/latest/src/egui/menu.rs.html#66).open_menu
    }
}

//////// [...] THEN LATER DOWN:

/// Stores the state for the context menu.
#[derive(Clone, Default)]
pub struct [MenuRootManager](https://docs.rs/egui/latest/egui/menu/struct.MenuRootManager.html) {
    inner: Option<[MenuRoot](https://docs.rs/egui/latest/src/egui/menu.rs.html#338-341)>,
}

impl [MenuRootManager](https://docs.rs/egui/latest/src/egui/menu.rs.html#293-295) {
    /// Show a menu at pointer if right-clicked response.
    ///
    /// Should be called from [`Context`] on a [`Response`]
    pub fn show<R>(
        &mut self,
        button: &[Response](https://docs.rs/egui/latest/src/egui/response.rs.html#22-136),
        add_contents: impl FnOnce(&mut [Ui](https://docs.rs/egui/latest/src/egui/ui.rs.html#51-113)) -> R,
    ) -> Option<InnerResponse<R>> {
        if let Some(root) = [self](https://docs.rs/egui/latest/src/egui/menu.rs.html#302).inner.[as_mut](https://doc.rust-lang.org/nightly/core/option/enum.Option.html#method.as_mut)() {
            let (menu_response, inner_response) = [root](https://docs.rs/egui/latest/src/egui/menu.rs.html#306).[show](https://docs.rs/egui/latest/src/egui/menu.rs.html#351-372)([button](https://docs.rs/egui/latest/src/egui/menu.rs.html#303), [add_contents](https://docs.rs/egui/latest/src/egui/menu.rs.html#304));
            if menu_response.is_close() {
                [self](https://docs.rs/egui/latest/src/egui/menu.rs.html#302).inner = None;
            }
            [inner_response](https://docs.rs/egui/latest/src/egui/menu.rs.html#307)
        } else {
            None
        }
    }

    fn is_menu_open(&self, id: [Id](https://docs.rs/egui/latest/src/egui/id.rs.html#34)) -> [bool](https://doc.rust-lang.org/nightly/std/primitive.bool.html) {
        [self](https://docs.rs/egui/latest/src/egui/menu.rs.html#317).inner.[as_ref](https://doc.rust-lang.org/nightly/core/option/enum.Option.html#method.as_ref)().[map](https://doc.rust-lang.org/nightly/core/option/enum.Option.html#method.map)(|m| [m](https://docs.rs/egui/latest/src/egui/menu.rs.html#318).id) == Some([id](https://docs.rs/egui/latest/src/egui/menu.rs.html#317))
    }
}

impl [std::ops::Deref](https://doc.rust-lang.org/nightly/core/ops/deref/trait.Deref.html) for [MenuRootManager](https://docs.rs/egui/latest/src/egui/menu.rs.html#293-295) {
    type Target = Option<[MenuRoot](https://docs.rs/egui/latest/src/egui/menu.rs.html#338-341)>;

    fn deref(&self) -> &Self::Target {
        &[self](https://docs.rs/egui/latest/src/egui/menu.rs.html#325).inner
    }
}

```

... and the [tokio](https://github.com/tokio-rs/tokio/blob/2c14f88c90947b5fd845892b3d968f7b342db04e/tokio/src/runtime/context/current.rs) link has no interdependencies, but [this](https://github.com/tokio-rs/tokio/blob/2c14f88c90947b5fd845892b3d968f7b342db04e/tokio/src/runtime/builder.rs) other one has:
```rust

pub struct Builder {
    /// Runtime type
    kind: Kind,

    /// Whether or not to enable the I/O driver
    enable_io: bool,
    nevents: usize,

    /// Whether or not to enable the time driver
    enable_time: bool,

    /// Whether or not the clock should start paused.
    start_paused: bool,

    /// The number of worker threads, used by Runtime.
    ///
    /// Only used when not using the current-

/// [...] AND LATER DOWN:

pub(crate) type ThreadNameFn = std::sync::Arc<dyn Fn() -> String + Send + Sync + 'static>;

#[derive(Clone, Copy)]
pub(crate) enum Kind {
    CurrentThread,
    #[cfg(feature = "rt-multi-thread")]
    MultiThread,
    #[cfg(all(tokio_unstable, feature = "rt-multi-thread"))]
    MultiThreadAlt,
}

impl Builder {
    /// Returns a new builder with the current thread scheduler selected.
    ///
    /// Configuration methods can be chained on the return value.
    ///
    /// To spawn non-`Send` tasks on the resulting runtime, combine it with a
    /// [`LocalSet`].

```

```

Reply:
![image](https://github.com/user-attachments/assets/59f030c0-08c7-41b9-b63f-ba3a600d59ad)

Look at the lines:

![image](https://github.com/user-attachments/assets/317e8e21-c862-4cd9-948e-7eef67924b26)
![image](https://github.com/user-attachments/assets/28ed5e7f-2c29-432f-a986-648adb11f2e3)

Reply:
> In the case of keeping all decl first while using top-to-down:
> 
> ```rust
> use std::sync::{Arc, Mutex};
> 
> lazy_static!{
>     static ref AAAA: A: A {...}
> }
> 
> struct A {b: Arc<Mutex<<B>>}
> 
> struct B {c: Vec<C>}
> 
> struct C {...}
> 
> impl A {...}
> 
> impl B {...}
> 
> impl C {...}
> 
> fn freefunction() {...}
> ```

Again:

struct C is used by B, so B comes first.
struct B is used by A, so A comes first.

this is top-to-down and allow us to first see the bigger abstractions, to later down see the details

Reply:
I write everything top-to-bottom as it is my (and most others) natural flow of thinking, as it is the way we tell stories and explain almost anything in life. I will not spend any more time with this bikeshedding.

Reply:
done.